batch.py

```diff
@@ -0,0 +1,761 @@
+#!/usr/bin/env python3
+# pylint: disable=wrong-import-position,import-error
+
+"""
+Hail Batch workflow for the rare-variant association analysis, including:
+
+- get relevant variants around a gene and export genotypes as plink files
+- generate input files for association tests 
+- run association tests
+"""
+
+# import python modules # is there a specific order, rationale for grouping imports??
+import os
+
+import click
+import logging
+
+from cloudpathlib import AnyPath
+from cpg_utils import to_path
+from cpg_utils.hail_batch import (
+    dataset_path,
+    get_config,
+    init_batch,
+    output_path,
+    remote_tmpdir,
+)
+
+import numpy as np
+import pandas as pd
+import xarray as xr  
+
+from limix.qc import quantile_gaussianize
+from numpy import eye, ones
+from pandas_plink import read_plink1_bin
+from scipy.stats import shapiro
+
+# qvalue imports
+import scipy as sp
+from scipy import interpolate
+
+import hail as hl
+import hailtop.batch as hb
+from hail.methods import export_plink
+
+from cellregmap import (  # figure out how to import this from github
+    run_gene_set_association,
+    run_burden_association,
+    omnibus_set_association,
+)
+
+# use logging to print statements, display at info level
+logging.basicConfig(
+    format="%(levelname)s:%(message)s", level=logging.INFO
+)  # consider removing some print statements?
+
+DEFAULT_JOINT_CALL_MT = dataset_path("mt/v7.mt")
+DEFAULT_ANNOTATION_HT = dataset_path("tob_wgs_vep/104/vep104.3_GRCh38.ht")  # atm VEP only
+
+# region SUBSET_VARIANTS
+
+
+def filter_variants(
+    mt_path: str,
+    samples: list[str],
+    output_mt_path: str,
+):
+    """Subset hail matrix table
+
+    Input:
+    joint call hail matrix table
+    set of samples for which we have scRNA-seq data
+
+    Output:
+    subsetted hail matrix table, containing only variants that:
+    1) are not ref-only, 2) biallelic, 3) meet QC filters, 4) are rare (MAF<5%)
+    """
+    # read hail matrix table object (WGS data)
+    init_batch()
+    mt = hl.read_matrix_table(mt_path)
+
+    # subset to relevant samples (samples we have scRNA-seq data for)
+    mt = mt.filter_cols(samples)  # figure out syntax
+
+    # densify (can this be done at the end?)
```

Comment:
I think one of the issues around densification was the number of alleles - from memory you needed to densify before you can reliably filter to biallelic variants, but I might be wrong. 

While the rest of this process is being worked on you could run this method in isolation, and then the rest of the code would operate on a smaller, normalised matrixtable. That way you can check the contents/structure, and make sure it's fit for purpose before the massive runtime part